### PR TITLE
feat: remove docker image when component deleted (#1111)

### PR DESCRIPTION
**Pull request by leoSRM**
**Issue #1111:**

**Description of changes:**
When a component which use docker image has been removed, check whether this image being used by other running components，if not, delete this image with docker cli commend rmi.

**Why is this change necessary:**
Before this change, unused images would be left and take lots of space.

**How was this change tested:**
- [x] Updated or added new unit tests.
- [x] Manual test.

**Any additional information or context required to review the change:**

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/integrationtests/java/com/aws/greengrass/integrationtests/componentmanager/ComponentManagerIntegTest.java [hunk 1]
```diff
@@ -23,6 +23,7 @@
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
+import org.mockito.Mock;
```
Comment (MikeDombo) [inline src/integrationtests/java/com/aws/greengrass/integrationtests/componentmanager/ComponentManagerIntegTest.java:26]: Remove unnecessary changes

File: src/main/java/com/aws/greengrass/componentmanager/builtins/ArtifactDownloader.java [hunk 2]
```diff
@@ -294,5 +295,12 @@ public boolean canSetFilePermissions() {
     public boolean canUnarchiveArtifact() {
         return true;
     }
+
+    /**
+     * Cleanup artifacts.
+     *
+     * @param componentStore componentStore
+     */
+    public abstract void cleanup(ComponentStore componentStore) throws Exception;
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/builtins/ArtifactDownloader.java:304]: I don't think this method should have any parameters. If the `ComponentStore` is needed to delete things, then that should be handled in the constructor of an instance of `ArtifactDownloader` instead.

This method should not throw a raw Exception. Create a new exception type, `PackageCleanupException` similar to `PackageDownloadException`

File: src/main/java/com/aws/greengrass/componentmanager/builtins/S3Downloader.java [hunk 3]
```diff
@@ -67,6 +68,11 @@ protected String getArtifactFilename() {
         return pathStrings[pathStrings.length - 1];
     }
 
+    @Override
+    public void cleanup(ComponentStore componentStore) throws Exception {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/builtins/S3Downloader.java:72]: This method should be implemented so it actually performs the cleanup. I think that the cleanup logic should be moved from the ComponentStore.

File: src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java [hunk 4]
```diff
@@ -223,9 +227,26 @@ ComponentRecipe getPackageRecipe(@NonNull ComponentIdentifier pkgId) throws Pack
...
+            ComponentRecipe recipe = getPackageRecipe(compId);
+            Path packageArtifactDirectory = resolveArtifactDirectoryPath(compId);
+            for (ComponentArtifact artifact : recipe.getArtifacts()) {
+                ArtifactDownloader downloader = artifactDownloaderFactory
+                        .getArtifactDownloader(compId, artifact, packageArtifactDirectory);
+                try {
+                    downloader.cleanup(this);
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java:247]: Absolutely do not re-throw the exception, this will cause wide-spread impact.

Look at how the existing logic handles exceptions. They will check

```
if (exception == null) {
                exception = e;
            } else {
                exception.addSuppressed(e);
            }
```

File: src/main/java/com/aws/greengrass/componentmanager/plugins/docker/exceptions/DockerImageDeleteException.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.componentmanager.plugins.docker.exceptions;
+
+import static com.aws.greengrass.deployment.errorcode.DeploymentErrorCode.DOCKER_RMI_ERROR;
+
+public class DockerImageDeleteException extends DockerImageDownloadException {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/plugins/docker/exceptions/DockerImageDeleteException.java:10]: I don't think that this should be a subclass of the download exception. A deletion exception is not related to downloading.

File: src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java [hunk 6]
```diff
@@ -223,15 +227,40 @@ ComponentRecipe getPackageRecipe(@NonNull ComponentIdentifier pkgId) throws Pack
      * @param compId component identifier
      * @throws PackageLoadingException if deletion of the component failed
      */
-    void deleteComponent(@NonNull ComponentIdentifier compId) throws PackageLoadingException {
+    void deleteComponent(@NonNull ComponentIdentifier compId, @NonNull ArtifactDownloaderFactory artifactDownloaderFactory) throws PackageLoadingException, InvalidArtifactUriException {
         logger.atDebug("delete-component-start").kv("componentIdentifier", compId).log();
-        IOException exception = null;
+        Exception exception = null;
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/ComponentStore.java:232]: only IOException should be thrown.

File: src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DockerImageDownloader.java [hunk 7]
```diff
@@ -276,4 +285,47 @@ private <T> T runWithConnectionErrorCheck(CrashableSupplier<T, Exception> task)
...
+    @Override
+    public void cleanup() throws PackageLoadingException {
+        // this docker image not only used by itself
+        if (!ifImageUsedByOther(componentStore)) {
+            try {
+                Image image = DockerImageArtifactParser
+                        .getImage(ComponentArtifact.builder().artifactUri(artifact.getArtifactUri()).build());
+                dockerClient.deleteImage(image);
+            } catch (InvalidArtifactUriException | DockerImageDeleteException e) {
+                logger.atWarn().kv("docker image", artifact.getArtifactUri())
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DockerImageDownloader.java:302]: this error should be thrown up to the caller and logged and handled by the caller.

File: src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DockerImageDownloader.java [hunk 8]
```diff
@@ -276,4 +285,41 @@ private <T> T runWithConnectionErrorCheck(CrashableSupplier<T, Exception> task)
...
+     * @return true: this image used by other; false: not used.
+     * @throws PackageLoadingException from getPackageRecipe
+     */
+    public boolean ifImageUsedByOther(ComponentStore componentStore) throws PackageLoadingException {
+        Map<String, Set<String>> allVersions = componentStore.listAvailableComponentVersions();
+        for (Map.Entry<String, Set<String>> versions : allVersions.entrySet()) {
+            String compName = versions.getKey();
+            Set<String> localVersions = new HashSet<>(versions.getValue());
+            for (String compVersion : localVersions) {
+                ComponentIdentifier identifier = new ComponentIdentifier(compName, new Semver(compVersion));
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DockerImageDownloader.java:314]: `new Semver` can throw an exception. Handle the exceptions that can happen here by logging, and continuing the loop.

File: src/main/java/com/aws/greengrass/componentmanager/builtins/ArtifactDownloader.java [hunk 9]
```diff
@@ -294,5 +297,11 @@ public boolean canSetFilePermissions() {
     public boolean canUnarchiveArtifact() {
         return true;
     }
+
+    /**
+     * Cleanup artifacts.
+     *
+     */
+    public abstract void cleanup() throws Exception;
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/builtins/ArtifactDownloader.java:305]: The throw here should be an `IOException` only.

**Comment by MikeDombo** [comment]
Please ensure that all the build and tests are passing. You have formatting issues which cause the build to fail.

File: src/main/java/com/aws/greengrass/componentmanager/plugins/docker/exceptions/DockerImageDeleteException.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.componentmanager.plugins.docker.exceptions;
+
+import com.aws.greengrass.deployment.exceptions.DeploymentException;
+
+import static com.aws.greengrass.deployment.errorcode.DeploymentErrorCode.DOCKER_RMI_ERROR;
+
+public class DockerImageDeleteException extends DeploymentException {
```
Comment (junfuchen99) [inline src/main/java/com/aws/greengrass/componentmanager/plugins/docker/exceptions/DockerImageDeleteException.java:12]: extends `PackageLoadingException` instead of DeploymentException

File: src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCode.java [hunk 11]
```diff
@@ -94,6 +94,7 @@ public enum DeploymentErrorCode {
     DOCKER_LOGIN_ERROR(DeploymentErrorType.DEPENDENCY_ERROR),
     DOCKER_PULL_ERROR(DeploymentErrorType.DEPENDENCY_ERROR),
     DOCKER_IMAGE_NOT_VALID(DeploymentErrorType.DEPENDENCY_ERROR),
+    DOCKER_RMI_ERROR(DeploymentErrorType.DEPENDENCY_ERROR),
```
Comment (junfuchen99) [inline src/main/java/com/aws/greengrass/deployment/errorcode/DeploymentErrorCode.java:97]: you don't need this error code because failure to remove docker image should not fail a deployment

**Comment by leoSRM** [comment]
> 

all right

File: src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DockerImageDownloader.java [hunk 12]
```diff
@@ -276,4 +286,50 @@ private <T> T runWithConnectionErrorCheck(CrashableSupplier<T, Exception> task)
...
+     */
+    public boolean ifImageUsedByOther(ComponentStore componentStore) throws PackageLoadingException {
+        Map<String, Set<String>> allVersions = componentStore.listAvailableComponentVersions();
+        for (Map.Entry<String, Set<String>> versions : allVersions.entrySet()) {
+            String compName = versions.getKey();
+            Set<String> localVersions = new HashSet<>(versions.getValue());
+            for (String compVersion : localVersions) {
+                try {
+                    ComponentIdentifier identifier = new ComponentIdentifier(compName, new Semver(compVersion));
+                    if (ObjectUtils.notEqual(identifier, this.identifier)) {
```
Comment (junfuchen99) [inline src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DockerImageDownloader.java:321]: Do not use `ObjectUtils` to compare since already have `compareTo` defined for ComponentIdentifier

File: src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DockerImageDownloader.java [hunk 13]
```diff
@@ -276,4 +286,50 @@ private <T> T runWithConnectionErrorCheck(CrashableSupplier<T, Exception> task)
...
+                        ComponentRecipe recipe = componentStore.getPackageRecipe(identifier);
+                        if (recipe.getArtifacts().stream().anyMatch(i -> i.getArtifactUri().equals(artifact.getArtifactUri()))) {
+                            return true;
+                        }
+                    }
+                } catch (SemverException e) {
+                    logger.atWarn().kv("identifier", identifier.getName()).setCause(e)
+                            .log("Error happened when semver being created.");
```
Comment (junfuchen99) [inline src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DockerImageDownloader.java:329]: remove period in the end of message. Also log `compVersion` in kv pair

File: src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DefaultDockerClient.java [hunk 14]
```diff
@@ -166,6 +167,23 @@ private Optional<UserNotAuthorizedForDockerException> checkUserAuthorizationErro
...
     }
 
+    /**
+     * Use docker command to delete the docker image
+     *
+     * @param image docker image to delete
+     * @throws DockerImageDeleteException if error is encountered
+     */
+    public void deleteImage(Image image) throws DockerImageDeleteException {
+        CliResponse response = runDockerCmd(String.format("docker rmi -f %s", image.getImageFullName()));
```
Comment (junfuchen99) [inline src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DefaultDockerClient.java:177]: Why do we need force deletion here?

Comment (leoSRM) [reply src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DefaultDockerClient.java:177]: > Why do we need force deletion here?

Incase the container which use this image not stoped, so you mean this pattern would not happen?

**Review by MikeDombo — DISMISSED** [review]

File: src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DefaultDockerClient.java (hunk 14 above)
Comment (leoSRM) [reply src/main/java/com/aws/greengrass/componentmanager/plugins/docker/DefaultDockerClient.java:177]: OK, will not use force deletion.

**Review by MikeDombo — DISMISSED** [review]

**Review by junfuchen99 — APPROVED** [review]

**Comment by indougnito** [comment]
Created #1448 to get around check failures

**Comment by junfuchen99** [comment]
Closing this pull request since it's being addressed in https://github.com/aws-greengrass/aws-greengrass-nucleus/pull/1448
